### DISCUSSION TITLE
Why the separate wrap/unwrap from the to/fromJSON methods

### DISCUSSION THREAD

**Discussion by bouk** [stephenh/ts-proto#864] [General]
Hello @stephenh!

One thing I don't understand about this project is why the wrap/unwrap methods are generated for some Messages like `google.protobuf.Value`. Why not just return the canonical JSON representation from the `to/fromJSON` methods and call that recursively when serializing?

**Comment by stephenh** [comment]
Hi @bouk ! I'm not sure actually... @boukeversteegh wrote most of our wrap/unwrap code ... I could poke around to refresh my memory, but as a lazy ask @boukeversteegh wdyt of @bouk 's question? :-) Thanks!

**Comment by boukeversteegh** [comment]
the core reason for needing some sort of conversion step, is that some protobuf messages have a json representation that is different from their structure. so we have 4 representations of the same message.

- canonical typescript representation: we have tried as much as possible to make the objects intuitive within typescript. such as optional values having `?` or `value | undefined` as a type.
- protobuf message format: this is the schema defined by protobuf. for example, a list is `message ListValue` which has a field `Values`. this is different from how we represent it in typescript, where it would just be an array, like `string[]`.
- json encoding representation: in most cases, this is pretty similar to the protobuf message format. for some well known types, there is a 'canonical representation'. such as the list mentioned before, or the 'Value' type, which in protobuf uses a "switch" to select the field, and there is one field for each type. But in JSON it's just represented as the representative type in json. so instead of sending `{ stringValue: "foobar"}` you send "foobar". note that not all (serverside) libraries do this, so you may want to be able to parse the original 1-to-1 json encoding of the protobuf message, just in case.
- protobuf binary encoding. here there are no smart representations of well known types. the values are encoded as the underlying well known type, following the original field numbers and structure.

by having a separate functions to convert a protobuf message either into typescript, or json, we can represent the object in all these ways.

- wrap/unwrap (not most clear name, admittedly) is to convert TS to the protobuf message schema, this is needed for binary encoding
- toJson/fromJson is needed when handling server generated messages. we need this because unset fields may be omitted, but protobuf requires that all fields are available on a protobuf message, whether they are set or not.

I remember there are more nuances to this that what I explained above. If you really want to get into it, I suggest trying out what you had in mind, and then running the tests, to see where things will break.

**Reply by stephenh** [reply]
Hi @boukeversteegh ! Thanks for the great reply! That was not as clear in my head as you laid out, so really appreciate that.

So, if I'm following, a tldr is that ts-proto _could_ generate types that match the the JSON encoding 1-for-1, but we've generally chosen "more idiomatic" TS than maybe what the JS spec says.

Actually, phrasing it that way, reminds me of our `useJsonWireFormat` option, which several users requested b/c it allows taking protobuf-generated JSON straight off the wire and just immediately casting it to the ts-proto-generated type, vs. going through a `fromJson` translation step. 

So perhaps that is what you're looking for?

Either way, @bouk does that answer your question?

**Reply by boukeversteegh** [reply]
You're welcome!

Let's look at a simple protobuf message to clarify the use of wrapping and json methods.

https://github.com/stephenh/ts-proto/blob/5ec618c64fa727eaf5dd08e8429900aca3375e22/integration/simple/simple.proto#L16-L34

## JSON conversion

`fromJSON` does a few things:

- Copy the JSON object's field values to the corresponding Typescript object fields. 👈 Trivial
- Convert JSON encoded values to Typescript native values (such as datetime and enums). 👈 This is the *idiomatic* part.
- Initialize missing fields to their default value (e.g. `""` for string, `0` for int64) 👈 This is required by protobuf specs.

https://github.com/stephenh/ts-proto/blob/5ec618c64fa727eaf5dd08e8429900aca3375e22/integration/simple/simple.ts#L508-L527


## Wrap

The converted `simple.proto` message does use any `wrap` methods, because it doesn't use well-known-types like ListValue or Value. `value.proto` has fields that are well known values, so before encoding, it needs to convert the Typescript representation (which is close to the canonical json format) to the underlying protobuf message structure.


https://github.com/stephenh/ts-proto/blob/5ec618c64fa727eaf5dd08e8429900aca3375e22/integration/value/value.proto#L6-L12

The `encode` method looks like:
https://github.com/stephenh/ts-proto/blob/5ec618c64fa727eaf5dd08e8429900aca3375e22/integration/value/value.ts#L21-L38

You can see that for all the 'special' fields (well known types), an additional 'wrap' step is needed. Here's `Value.wrap`

https://github.com/stephenh/ts-proto/blob/5ec618c64fa727eaf5dd08e8429900aca3375e22/integration/value/google/protobuf/struct.ts#L402-L420

Here you can see clearly that what is a simple variable in Typescript, will be converted into an object and assigned to the field for the type of that variable.

---

So in short:

- `json` methods are needed to make sure all implicit fields are initialized, and json values are mapped to typescript values.
 - `wrap` methods are needed to convert the well-known-types to and from their canonical representation
